### PR TITLE
29 google oauth

### PR DESCRIPTION
**Pull request by peter-tanner**
## Change Summary
- Adds google oauth as a login option
- Adds some basic frontend items like login button and redirect to get oauth working

### Change Form
**Fill this up (NA if not available). If a certain criteria is not met, can you please give a reason.**

- [x] The pull request title has an issue number
- [x] The change works by "Smoke testing" or quick testing
- [ ] The change has tests

Only tests for Django authentication. No tests for Google OAuth - too difficult

- [x] The change has documentation

## Other Information
**[Is there anything in particular in the review that I should be aware of?]**

### CODE REVIEW

File: backend/authentication/tests/test_authentication.py [hunk 1]
```diff
@@ -2,6 +2,10 @@
 from rest_framework import status
 
 
+def fail(setup_indeaa):
+    assert "True" == "False"
+
```
Comment (frinzekt) [inline backend/authentication/tests/test_authentication.py:7]: do you need this?

File: backend/.vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
+    "python.linting.flake8Enabled": true,
+    "python.linting.enabled": true,
+    "python.linting.flake8Args": ["--max-line-length=150", "--ignore=F405,F403"],
+}
```
Comment (frinzekt) [inline backend/.vscode/settings.json:5]: i suppose i am fine with this... but for usual case most devs have their own settings json (just saying)

File: backend/authentication/views.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
+import urllib
+
+from allauth.socialaccount.providers.google import views as google_views
+from allauth.socialaccount.providers.oauth2.client import OAuth2Client
+from dj_rest_auth.registration.views import SocialLoginView
+from django.urls import reverse
+from django.views.generic.base import RedirectView
```
Comment (frinzekt) [inline backend/authentication/views.py:7]: nice job finding a django view that is built in

File: docker/indeaa-local.env [hunk 4]
```diff
@@ -50,6 +50,12 @@ SUPERUSER_PASSWORD=Password123
 PGADMIN_DEFAULT_EMAIL=admin@admin.com
 PGADMIN_DEFAULT_PASSWORD=Password123
 
+# ======
+# GOOGLE
+# ======
+GOOGLE_CLIENT_ID=287885917936-040ee4p7ppe8krf1a0dh3nvaedm76qt1.apps.googleusercontent.com
+GOOGLE_SECRET=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
+
```
Comment (frinzekt) [inline docker/indeaa-local.env:58]: yeah i suppose i don't mind having these environment variables right over here... 
BUT KEEP IN MIND to make this different from production

File: frontend/pages/login.tsx [hunk 5]
```diff
@@ -67,59 +63,75 @@ const Login = (props: Props) => {
...
+            <Button
+              startIcon={<Google />}
+              variant="contained"
+              size="large"
+              type="submit"
+              sx={{ marginTop: MARGIN, marginBottom: MARGIN }}
+            >
+              Login with Google
+            </Button>
+          </form>
```
Comment (frinzekt) [inline frontend/pages/login.tsx:132]: i was actually thinking that the button is an `href` using an anchor tag. What do you think @MouseAndKeyboard ?

File: frontend/pages/auth/google.tsx [hunk 6]
```diff
@@ -0,0 +1,68 @@
+import AppContext from '@/components/Context/TopLevelContext';
```
Comment (frinzekt) [inline frontend/pages/auth/google.tsx:1]: i presume this is the page that receives the URL? based on reading the backend code.

looks good in my own opinion

**Comment by peter-tanner** [comment]
Ready for review I think

**Review by frinzekt — CHANGES_REQUESTED** [review]
Code and setup looks good (i haven't smoke tested it yet).

Can you do a short documentation on it? The important thing document is

0. overview of oauth (you can even just put links to the documentations you've used that found to be relevant)
1. where to find the backend configs
2. what part of the frontend receives the code

Note: you dont have to explain every single thing. Just direct the user to where they should see it

File: backend/authentication/tests/test_authentication.py (hunk 1 above)
Comment (peter-tanner) [reply backend/authentication/tests/test_authentication.py:7]: Whoops I do not need that I was trying to get tests running

**Comment by peter-tanner** [comment]
Yes I wrote some docs on how to get google oauth through google console. I'll add in the other stuff on oauth

File: frontend/pages/auth/google.tsx (hunk 6 above)
Comment (peter-tanner) [reply frontend/pages/auth/google.tsx:1]: Yep

File: docker/indeaa-local.env (hunk 4 above)
Comment (peter-tanner) [reply docker/indeaa-local.env:58]: Yep.
Should I put some placeholder text there instead? I have written in the docs that this needs to be changed before starting it

Comment (frinzekt) [reply docker/indeaa-local.env:58]: i think thats all good

**Comment by frinzekt** [comment]
> Yes I wrote some docs on how to get google oauth through google console. I'll add in the other stuff on oauth

yea i didnt see the docs... looking at it now, it looks good, but yeah, some other stuff to go with it will be good

**Review by frinzekt — COMMENTED** [review]
I am removing my "requested changes". @chayG5 can you review this code and smoke test it (make sure that it works), then i will leave the hand to you to approve

File: backend/.vscode/settings.json (hunk 2 above)
Comment (peter-tanner) [reply backend/.vscode/settings.json:5]: Removed `settings.json`, I figured out how to get the linting working i just needed another block for formatting if anyone else is reading

```json
"python.linting.enabled": true,
  "python.linting.flake8Enabled": true,
  "python.formatting.autopep8Args": [
    "--max-line-length=150",
    "--ignore=F405,F403"
  ],
  "python.linting.flake8Args": [
    "--max-line-length=150",
    "--ignore=F401"
  ],
  ```
  in your global `settings.json`

**Comment by peter-tanner** [comment]
If anyone wants to test it you need to create a web application on google cloud and create an oauth client id credential (Guide is in docs), or you can send me your google account email and try on the one I have been testing on, as I need to add test users manually.

**Comment by frinzekt** [comment]
> as I need to add test users manually.

just a question... How would that work in a production system? I also don't recall that any other email needs to be added to anywhere else?

**Comment by peter-tanner** [comment]
For production we need to get the app in production, I haven't tried yet but possibly need verification from Google?
I'm still using the app in testing mode

> In Production:
> Once you set your app status as "In production," your app will be available to anyone with a Google Account. Depending on how you configure your OAuth screen, you may have to [submit your app for verification ](https://support.google.com/cloud/answer/9110914).
> 
> Testing:
> If your app is still being tested and built, you can set your status to "testing". In this state, you can test your app with a limited number of users.

**Comment by frinzekt** [comment]
ahh okay sounds good, yeah i add my email there (i'll send you a pm)

**Comment by frinzekt** [comment]
I received an error smoke testing this
```
indeaa_django    | [2022-07-19 16:23:22,935] [97170f79a59f43bdb3b8bb73d39367c3] ERROR [authentication.views.GoogleLogin:87] [IndEAA_0.0.0] Anonymous User: Unknown Error Occurred: 'NoneType' object has no attribute 'is_authenticated' 
indeaa_django    | Traceback (most recent call last):
indeaa_django    |   File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 506, in dispatch
indeaa_django    |     response = handler(request, *args, **kwargs)
indeaa_django    |   File "/usr/local/lib/python3.9/site-packages/dj_rest_auth/views.py", line 124, in post
indeaa_django    |     self.serializer.is_valid(raise_exception=True)
indeaa_django    |   File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 227, in is_valid
indeaa_django    |     self._validated_data = self.run_validation(self.initial_data)
indeaa_django    |   File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 429, in run_validation
indeaa_django    |     value = self.validate(value)
indeaa_django    |   File "/usr/local/lib/python3.9/site-packages/dj_rest_auth/registration/serializers.py", line 151, in validate
indeaa_django    |     complete_social_login(request, login)
indeaa_django    |   File "/usr/local/lib/python3.9/site-packages/allauth/socialaccount/helpers.py", line 156, in complete_social_login
indeaa_django    |     return _complete_social_login(request, sociallogin)
indeaa_django    |   File "/usr/local/lib/python3.9/site-packages/allauth/socialaccount/helpers.py", line 167, in _complete_social_login
indeaa_django    |     if request.user.is_authenticated:
indeaa_django    | AttributeError: 'NoneType' object has no attribute 'is_authenticated'
indeaa_django    | [2022-07-19 16:23:22,943] [97170f79a59f43bdb3b8bb73d39367c3] INFO [authentication.views.GoogleLogin:60] [IndEAA_0.0.0] Anonymous User: Sent 500 POST /api/v1/authentication/google/  
indeaa_django    | [IndEAA_0.0.0] Anonymous User: Internal Server Error: /api/v1/authentication/google/ 
indeaa_django    | [2022-07-19 16:23:22,943] [97170f79a59f43bdb3b8bb73d39367c3] ERROR [django.request:224] [IndEAA_0.0.0] Anonymous User: Internal Server Error: /api/v1/authentication/google/ 
indeaa_django    | [19/Jul/2022 16:23:22] [IndEAA_0.0.0] Anonymous User: "POST /api/v1/authentication/google/ HTTP/1.1" 500 118 
indeaa_django    | [2022-07-19 16:23:23,020] [425b9b4fd1f949f8a1bf6234485072db] ERROR [authentication.views.GoogleLogin:87] [IndEAA_0.0.0] Anonymous User: Unknown Error Occurred: 'NoneType' object has no attribute 'is_authenticated' 
indeaa_django    | Traceback (most recent call last):
indeaa_django    |   File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 506, in dispatch
indeaa_django    |     response = handler(request, *args, **kwargs)
indeaa_django    |   File "/usr/local/lib/python3.9/site-packages/dj_rest_auth/views.py", line 124, in post
indeaa_django    |     self.serializer.is_valid(raise_exception=True)
indeaa_django    |   File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 227, in is_valid
indeaa_django    |     self._validated_data = self.run_validation(self.initial_data)
indeaa_django    |   File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 429, in run_validation
indeaa_django    |     value = self.validate(value)
indeaa_django    |   File "/usr/local/lib/python3.9/site-packages/dj_rest_auth/registration/serializers.py", line 151, in validate
indeaa_django    |     complete_social_login(request, login)
indeaa_django    |   File "/usr/local/lib/python3.9/site-packages/allauth/socialaccount/helpers.py", line 156, in complete_social_login
indeaa_django    |     return _complete_social_login(request, sociallogin)
indeaa_django    |   File "/usr/local/lib/python3.9/site-packages/allauth/socialaccount/helpers.py", line 167, in _complete_social_login
indeaa_django    |     if request.user.is_authenticated:
indeaa_django    | AttributeError: 'NoneType' object has no attribute 'is_authenticated'
indeaa_django    | [2022-07-19 16:23:23,021] [425b9b4fd1f949f8a1bf6234485072db] INFO [authentication.views.GoogleLogin:60] [IndEAA_0.0.0] Anonymous User: Sent 500 POST /api/v1/authentication/google/  
indeaa_django    | [IndEAA_0.0.0] Anonymous User: Internal Server Error: /api/v1/authentication/google/ 
indeaa_django    | [2022-07-19 16:23:23,021] [425b9b4fd1f949f8a1bf6234485072db] ERROR [django.request:224] [IndEAA_0.0.0] Anonymous User: Internal Server Error: /api/v1/authentication/google/ 
indeaa_django    | [19/Jul/2022 16:23:23] [IndEAA_0.0.0] Anonymous User: "POST /api/v1/authentication/google/ HTTP/1.1" 500 118 
```

**Comment by peter-tanner** [comment]
Ah. I think this is a bug with allauth or I am not using it right. Could not find anything about this error online so I'm not too sure.
I fixed this a while ago by changing that line 167 from
```py
if request.user.is_authenticated:
```
to
```py
if request.user and request.user.is_authenticated:```

**Comment by frinzekt** [comment]
yeah the temporary fix that you had should not be applied. That was just temporary for development... it needs to be taken a look at

**Comment by frinzekt** [comment]
Note you are not the only one experiencing the same issue like this...
see

https://stackoverflow.com/questions/61643044/django-rest-framework-getting-error-nonetype-object-has-no-attribute
https://github.com/encode/django-rest-framework/issues/7618
https://github.com/jpadilla/django-rest-framework-jwt/issues/444

we have something similar see `IndEAAAnonRateThrottle`

**Comment by frinzekt** [comment]
this looks good in my opnion so I am merging this in now

**Pull request merged**